docs/ALU Design.md

```diff
@@ -0,0 +1,758 @@
+- there is no automatic wraparound as in the case of `as`.
+
+Should type conversion be necessary in the implementation of the operations, it will need to handle
+the type conversion with `try_into()` and manual recover from errors:
+
+```rust
+let result: u32 = match sum.try_into() {
+  Ok(val) => val,
+  Err(_) => {
+    // Handle the wraparound manually
+  }
+};
+```
+
+### Statefulness
+
+A real Arithmetic-Logic Unit in a CPU is a finite state machine. Some states, interesting from the
+programmer's point of view, can be captured as contents of the CPU registers. Such state is e.g. the
+next instruction (as pointed to by Program Counter or its equivalent), values of operands stored in
+two general purpose registers or the result of the last operation stored in another GP register and
+a flag register, where specific bits signal certain conditions (e.g. the result being zero or an
+integer overflow).
+
+The LLVM-to-Cairo infrastructure needs to deliver pieces of code translating generic LLVM arithmetic
+operations to their counterparts specific to the Cairo VM architecture. This translation will be
+done on the code level, during one of the LLVM-to-Cairo pipeline stages. Namely, this will be not
+runtime translation, but rather a compilation time one. Therefore, there is no global state to be
+managed during that time.
+
+Additionally, it has been noticed
+[in one of the experiments](https://github.com/reilabs/llvm-to-cairo/issues/27#issuecomment-2391893640),
+that LLVM IR follows the same principle of not managing the internal state of arithmetic operations.
+This is either done by
+
+- returning a tuple containing both the operation result and the state information:
+
+```llvm
+%0 = call { i64, i1 } @llvm.uadd.with.overflow.i64(i64 %left, i64 %right), !dbg !17
+%_3.0 = extractvalue { i64, i1 } %0, 0, !dbg !17
+%_3.1 = extractvalue { i64, i1 } %0, 1, !dbg !17
+br i1 %_3.1, label %panic, label %bb1, !dbg !17
+```
+
+- by demanding that input data is correct and producing undefined behavior otherwise,
+- by emitting poison values, if producing a correct value is not possible.
+
+Based on these observations, we decide to deliver the ALU as a collection of stateless arithmetic
+operations.
+
+### Tests
+
+Cairo has an
+[integrated test framework](https://book.cairo-lang.org/ch10-01-how-to-write-tests.html), similar to
+the one offered by Rust. Our ALU implementation will then come with a test suite to verify that we
+implement the desired behavior, e.g. to make sure we indicate overflow on obvious situations like
+`0xFFFFFFFF + 1` for a `u32` add.
+
+## Design
+
+### Overview
+
+The ALU will be implemented as a source code written in
+[Cairo](https://book.cairo-lang.org/title-page.html). During the
+[LLVM-to-Cairo compilation pipeline](https://www.notion.so/reilabs/System-Architecture-113d2f80c874802b8480d997347933a2?pvs=4)
+it will be translated to `FlatLowered` objects. Then, on the linking phase, arithmetic operations
```

Comment:
Not quite! We will have a modified version of the existing cairo compiler that will take our polyfill implementations, compile them to `FlatLowered` and then extracted to `.flo` files. The linker will then take _all_ the `.flo` files, those from the LLVM to Cairo compiler **and** the modified Cairo compiler and link them together.